[exporter/fileexporter] Support for rotation of telemetry files. (#14096)
open-telemetry/opentelemetry-collector-contrib: exporter/fileexporter/README.md

# File Exporter

| Status                   |                       |
| ------------------------ | --------------------- |
| Stability                | [alpha]               |
| Supported pipeline types | traces, metrics, logs |
| Distributions            | [core], [contrib]     |

Exporter supports the following features：

+ Support for writing pipeline data to a JSON file.
+ Support for rotation of telemetry files.

The data is written in.[Protobuf JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json). using [OpenTelemetry protocol](https://github.com/open-telemetry/opentelemetry-proto).

Please note that there is no guarantee that exact field names will remain stable.
This intended for primarily for debugging Collector without setting up backends.

## Getting Started

The following settings are required:
- `path` (no default): where to write information.

The following settings are optional:
- `rotation` settings to rotate telemetry files.

>     - max_size:  [default: 100]: the maximum size in megabytes of the telemetry file before it is rotated.

Reviewer: I think since we are stuck with v2 for now, we should call this `max_megabytes`. When v3 is available, we can add a new `max_size` field and deprecate this one.

Author: > Thanks for the quick updates. A couple more things. Apologies for marking up the docs so much. What you have is good but I'm being picky.

Your suggestions are very good, I can learn a lot from your suggestion. Thank you again~

Reviewer: This one still needs `max_size` -> `max_megabytes`

    - max_days: [no default (unlimited)]: the maximum number of days to retain telemetry files based on the timestamp encoded in their filename.
    - max_backups: [default: 100]: the maximum number of old telemetry files to retain.
    - localtime : [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.

> ## How to name rotated files 

Reviewer: ```suggestion
## File Rotation
```

> File opens or creates the file on first Write. If the file exists and is less than MaxSize megabytes, fileexporter will open and append to that file.

Reviewer: ```suggestion
Telemetry is first written to a file that exactly matches the `path` setting. When the file size exceeds `max_megabytes` or age exceeds `max_days`, the file will be rotated.
```

> If the file exists and its size is >= MaxSize megabytes,**the file is renamed by putting the current time in a timestamp** 

Reviewer: ```suggestion
When a file is rotated, **it is renamed by putting the current time in a timestamp** 
```

in the name immediately before the file's extension (or the end of the filename if there's no extension).

> **A new telemetry file will be created using original filename.**

Reviewer: ```suggestion
**A new telemetry file will be created at the original `path`.**
```

> For example,  if your filename is `data.json`, when the file exists and its size is >= MaxSize megabytes, this file will be renamed to `data-2022-09-14T05-02-14.173.json`, and a new telemetry file created with `data.json`.

Reviewer: ```suggestion
For example, if your `path` is `data.json` and rotation is triggered, this file will be renamed to `data-2022-09-14T05-02-14.173.json`, and a new telemetry file created with `data.json`.
```

The following settings are required:

- `path` (no default): where to write information.
- `rolling_logger_options` (default): an option to self-rotate log files.

>     - max_size:   the maximum size in megabytes of the log file before it gets rotated. It defaults to 100 megabytes.

Reviewer 2: How about using strings and parsing them as byte sizes, like with https://pkg.go.dev/github.com/alecthomas/units?

>     - max_size:   the maximum size in megabytes of the log file before it gets rotated. It defaults to 100 megabytes.

Reviewer: I had the same thought, but the underlying library does not allow it: https://github.com/natefinch/lumberjack/issues/37

It looks like it may be added in the next major version, so perhaps then we can implement `max_size` as you've suggested (and still support `max_megabytes` if necessary).

>     - max_age:   the maximum number of days to retain old log files based on the  timestamp encoded in their filename.The default is not to remove old log files based on age.

Reviewer: > based on the timestamp encoded in their filename

We will need a brief explanation of how rotated files are named.

>     - max_age:   the maximum number of days to retain old log files based on the  timestamp encoded in their filename.The default is not to remove old log files based on age.

Author: > > based on the timestamp encoded in their filename
> 
> We will need a brief explanation of how rotated files are named.

> 

Thank you，I will describe how rotated files are named～

>     - max_backups:  the maximum number of old log files to retain.  The default is to retain all old log files

Reviewer: ```suggestion
    - max_backups: [no default (unlimited)]: the maximum number of old log files to retain.
```

Reviewer: I think possibly we should include a default value here, something high enough that it does not easily affect users, but low enough that it will not crash most hosts. Maybe 100, so we have effectively a max of 10GB for all files. What do you think?

Author: I think It's a great idea.

>     - localtime : determines if the time used for formatting the timestamps in backup files is the computer's local time.  The default is to use UTC time.

Reviewer: ```suggestion
    - localtime: [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.
```
